lib/common/zstd_internal.h

```diff
@@ -227,7 +227,9 @@ void ZSTD_wildcopy(void* dst, const void* src, ptrdiff_t length, ZSTD_overlap_e
         COPY16(op, ip);
       }
       while (op < oend - 8);
-      COPY8(op, ip);
+
+      if (op < oend)
```

Comment:
This new branch _might_ affect performance, and is probably worth a new decompression speed measurement.

Reply:
I agree

Reply:
It looks like it's somewhere around 0.5% slower with the branch, in other words about 7% faster overall.

Reply:
Not sure if you've tried this, but would it be possible to fold this COPY8() into the first `if`? Something like: if `length & 15 < 8`? Then we should be able to guarantee that the COPY16() loop finishes the job and only overruns `8`.

Reply:
https://gcc.godbolt.org/z/MTe_1Y is the generated code for gcc.

```c
if ((length & 8) == 0)
    COPY8(op, ip);
do
    COPY16(op, ip);
while (op < oend);
```

I did some initial testing, and it seems like it might be a bit faster.

Reply:
Even better and less contrived:

```c
while (op < oend - 8)
    COPY16(op, ip);
COPY8(op, ip);
```

Reply:
What if `length == 17` ?

Reply:
Comparing gcc-9.1 and gcc-8.0 on the 3 versions (buggy, yours, and mine) with `zstd -b1 silesia.tar` I get:

| compiler | mgrice | terrelln | buggy |
| --- | --- | --- | --- |
| gcc | 1438.8 MB/s | 1443.2 MB/s | 1456.8 MB/s |
| clang | 1196.6 MB/s | 1216.6 MB/s | 1199.7 MB/s |